sklearn/utils/isfinite.pyx

```diff
@@ -0,0 +1,152 @@
+# Author: jakirkham
+
+cimport libc
+cimport libc.math
+from libc.math cimport isfinite as _isfinite, isinf as _isinf
+from libc.stdint cimport uint16_t
+
+cimport cython
+
+cimport numpy
+import numpy
+numpy.import_array()
+
+
+cdef extern from "numpy/halffloat.h":
+    ctypedef uint16_t npy_half
+
+    # conversion function
+    float npy_half_to_float(npy_half h);
+
+
+cdef fused fprecision:
+    npy_half
+    float
+    double
```

Comment:
I suggest only focusing on floats and doubles since these are the supported dtypes for floating point values in scikit-learn. All other dtypes paths can go down the original implementation.